// Copyright (c) 2018 IoTeX
// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
// License 2.0 that can be found in the LICENSE file.

syntax = "proto3";

package simulator;

// The simulator service definition.
service Simulator {
  rpc Ping (Request)		returns (stream Reply) {}
  rpc Init (InitRequest)	returns (stream Proposal) {}
  rpc Exit (Empty)              returns (Empty) {}
}

// The request message containing the playerID and the value of the message being fed into the consensus engine
message Request {
  int32 recipient	 = 1;
  int32 sender           = 2;
  uint32 internalMsgType = 3;
  string value		 = 4;
}

// The request message telling the server to initialize the necessary parameters
message InitRequest {
  int32 nHonest      = 1;
  int32 nFS	     = 2;
  int32 nBF	     = 3;
  int32 nConnections = 4;

  // trick to pass 2D arrays through gRPC
  message Connection {
    repeated int32 nodes = 1;
  }
  repeated Connection connections = 5;
}

// The response message returning the output of the consensus engine
message Reply {
  int32  messageType	 = 1;
  uint32 internalMsgType = 2;
  string value		 = 3;
  int32  recipient       = 4;
}

message Proposal {
  int32 playerID	 = 1;
  uint32 internalMsgType = 2;
  string value		 = 3;
}

// an empty message
message Empty {}

